### DISCUSSION TITLE
Can xrdp do TurboVNC, or only TigerVNC?

### DISCUSSION THREAD

**Discussion by dstromberg** [neutrinolabs/xrdp#2116] [General]
I have a need for fast graphics with sound - from a macOS laptop at my home to a CentOS 7 machine in a remote datacenter. Specifically, I really want to run Cinnamon on the CentOS system, sometimes with Youtube videos and Zoom.

TigerVNC is too slow and gives no sound. TurboVNC is almost comfortably fast, but still gives no sound. xrdp+TigerVNC gives sound, but TigerVNC seems to be slowing it down.

I tried upgrading my home internet plan, but AT&T could only muster an additional 25Mbps down and no change up. They're looking into doing better though.

Is there a way of building/configuring xrdp to use TurboVNC instead of TigerVNC? It seems like that combo might work for me.

I'm asking in part because xrdp, at least as it appears in the CentOS repos, seems to require TigerVNC.

Thanks!

**Comment by matt335672** [comment]
Hi @dstromberg 

The speed of your VNC server is irrelevant if you're using xrdp with the Xvnc backend, as the VNC data stays within the local machine:-

```
                             +-------------------------------------------+
                             |                                           |
+-------------+              |   +--------------+             +------+   |
| XRDP client | <------------|-->| xrdp process |  <--------- | Xvnc |   |
+-------------+   [RDP data] |   +--------------+  [VNC data] +------+   |
                             |                                           |
                             +-------------------------------------------+
```

The VNC data is uncompressed. The xrdp process converts the screen changes to RDP screen changes using appropriate compression for the protocol, and forwards them to the XRDP client.

By all means try a different Xvnc server. You just have to install it and then make sure that `/etc/xrdp/sesman.ini` starts the correct VNC server.

A better option for you might be to install `xorgxrdp` on the CentOS 7 machine, and connect using the `Xorg` backend rather than the `Xvnc` backend. This may give you better results, as in this configuration the X server and xrdp communicate using shared memory.

IIRC, CentOS does not use `xorgxrdp` as the default as there is far more coupling with the X server internals, and this causes breakage when the X server version is changed. However, for CentOS 7 this will not happen as RHEL 7 is now in Maintenance Support 2.

**Comment by dstromberg** [comment]
On Wed, Jan 12, 2022 at 2:00 AM matt335672 ***@***.***> wrote:

> Hi @dstromberg <https://github.com/dstromberg>
>
> The speed of your VNC server is irrelevant if you're using xrdp with the
> Xvnc backend, as the VNC data stays within the local machine:-
>
>                              +-------------------------------------------+
>                              |                                           |
> +-------------+              |   +--------------+             +------+   |
> | XRDP client | <------------|-->| xrdp process |  <--------- | Xvnc |   |
> +-------------+   [RDP data] |   +--------------+  [VNC data] +------+   |
>                              |                                           |
>                              +-------------------------------------------+
>
> I see.

> A better option for you might be to install xorgxrdp on the CentOS 7
> machine, and connect using the Xorg backend rather than the Xvnc backend.
> This may give you better results, as in this configuration the X server and
> xrdp communicate using shared memory.
>
Does xorgxrdp require a physical framebuffer?  The CentOS machine is
running headless.

I tried
https://www.ilhamsp.com/2021/01/remote-desktop-connection-centos-7.html ,
except I installed xorgxrdp manually (still using yum though), used
iptables instead of firewall-cmd, and I didn't install TightVNC - instead
going with the recommended TigerVNC.  Also, I didn't make any SELinux
changes.

But the communication was still pretty choppy. Unusably so, I guess.  I
didn't reboot.

Do I need to do anything further to get xorgxrdp to kick in?

If there's a fine manual I should be reading, feel free to point me at it.

Thanks!

**Reply by matt335672** [reply]
On C7, the xorgxrdp interface is commented out on installation, as it isn't the default.

Edit `/etc/xrdp/xrdp.ini`, and find these lines:-

```
#[Xorg]
#name=Xorg
#lib=libxup.so
#username=ask
#password=ask
#ip=127.0.0.1
#port=-1
#code=20
```

Uncomment these lines, and save the file.

Now restart xrdp:-

```
sudo systemctl restart xrdp
```

When you connect, you will now be able to select a session type of Xorg:-
![image](https://user-images.githubusercontent.com/30179339/149310418-6b854351-e857-404e-ba81-156a54daa3a4.png)

that will start a session with xorgxrdp.

You can check this with a ps -ef:-

<pre>
$ ps -ef | grep ' X'
testuser  2454  2451  3 10:14 ?        00:00:01 <b>Xorg</b> :10 -auth .Xauthority -config xrdp/xorg.conf -noreset -nolisten tcp -logfile .xorgxrdp.%s.log
testuser  3571  3523  0 10:15 pts/1    00:00:00 grep --color=auto  X
</pre>

you'll have `Xvnc` here if you're running with Xvnc.

**Reply by dstromberg** [reply]
I appear to be on Xorg:
$ ps -ef | grep ' X'
cmd output started 2022 Thu Jan 13 08:35:44 AM PST
dastrom+  24142  24139  0 08:33 ?        00:00:01 Xorg :10 -auth .Xauthority -config xrdp/xorg.conf -noreset -nolisten tcp -logfile .xorgxrdp.%s.log
root      25264  21464  0 08:35 pts/1    00:00:00 grep  X

I'm not seeing the option to start with Xorg - the "Login to" window you mention above doesn't seem to be coming up for me, but perhaps I don't need it given that ps|grep.

I manually killed xrdp and start'd it with systemctl restart xrdp.  I also logged out and back in.

Xorg might be a little faster, but still pretty unusable for Youtube (and presumably, Zoom as well).  It's still not as fast as TurboVNC, but as a reminder: TurboVNC wasn't giving sound.

Any further suggestions?

**Reply by matt335672** [reply]
If you're running on Xorg, set your client colour depth to 24 or 32, and specify a 'LAN connection' in the client, even if you're not on a LAN. That should enable the RFX codec which might give you better results. I don't know how you'd do that on the MAC client, but it's a simple dropdown on the default Windows client.

If that's not good enough, we can't offer you anything better at the moment. Stick with TurboVNC and investigate tunnelling the sound over the network using another protocol. I can't advise you on that I'm afraid.
